drivers/i2c/i2c_numaker.c

```diff
@@ -0,0 +1,808 @@
+ * 3. Force I2C Stop to end the transfer (thread)
+ *
+ * Passive transfer (slave only):
+ * 1. Prepare callback (thread)
+ * 2. Do data transfer via above callback (ISR)
+ */
+
+static void m_numaker_i2c_master_xfer_msg_read_next_byte(const struct device *dev);
+static void m_numaker_i2c_master_xfer_msg_read_last_byte(const struct device *dev);
+static void m_numaker_i2c_master_xfer_msg_end(const struct device *dev);
+static void m_numaker_i2c_master_xfer_end(const struct device *dev, bool do_stop);
+
+struct i2c_numaker_config {
+	I2C_T *i2c_base;
+	const struct reset_dt_spec reset;
+	uint32_t clk_modidx;
+	uint32_t clk_src;
+	uint32_t clk_div;
+	const struct device *clkctrl_dev;
+	uint32_t irq_n;
+	void (*irq_config_func)(const struct device *dev);
+	const struct pinctrl_dev_config *pincfg;
+	uint32_t bitrate;
+};
+
+struct i2c_numaker_data {
+	struct k_sem lock;
+	uint32_t dev_config;
+	/* Master transfer context */
+	struct {
+		struct k_sem xfer_sync;
+		uint16_t addr;
+		struct i2c_msg *msgs_beg;
+		struct i2c_msg *msgs_pos;
+		struct i2c_msg *msgs_end;
+		uint8_t *buf_beg;
+		uint8_t *buf_pos;
+		uint8_t *buf_end;
+	} master_xfer;
+#ifdef CONFIG_I2C_TARGET
+	/* Slave transfer context */
+	struct {
+		struct i2c_target_config *slave_config;
+		bool slave_addressed;
+	} slave_xfer;
+#endif
+};
+
+static int i2c_numaker_configure(const struct device *dev, uint32_t dev_config)
+{
+	const struct i2c_numaker_config *config = dev->config;
+	struct i2c_numaker_data *data = dev->data;
+	uint32_t bitrate;
+
+	/* Check address size */
+	if (dev_config & I2C_ADDR_10_BITS) {
+		LOG_ERR("10-bits address not supported");
+		return -ENOTSUP;
+	}
+
+	switch (I2C_SPEED_GET(dev_config)) {
+	case I2C_SPEED_STANDARD:
+		bitrate = KHZ(100);
+		break;
+	case I2C_SPEED_FAST:
+		bitrate = KHZ(400);
+		break;
+	case I2C_SPEED_FAST_PLUS:
+		bitrate = MHZ(1);
+		break;
+	default:
+		LOG_ERR("Speed code %d not supported", I2C_SPEED_GET(dev_config));
+		return -ENOTSUP;
+	}
+
+	I2C_T *i2c_base = config->i2c_base;
+	int err = 0;
+
+	k_sem_take(&data->lock, K_FOREVER);
+	irq_disable(config->irq_n);
+
+#ifdef CONFIG_I2C_TARGET
+	if (data->slave_xfer.slave_addressed) {
+		LOG_ERR("Reconfigure with slave being busy");
+		err = -EBUSY;
+		goto done;
+	}
+#endif
+
+	I2C_Open(i2c_base, bitrate);
+	/* INTEN bit and FSM control bits (STA, STO, SI, AA) are packed in one register CTL0. */
+	i2c_base->CTL0 |= (I2C_CTL0_INTEN_Msk | I2C_CTL0_I2CEN_Msk);
+	data->dev_config = dev_config;
+
+done:
+
+	irq_enable(config->irq_n);
+	k_sem_give(&data->lock);
+
+	return err;
+}
+
+static int i2c_numaker_get_config(const struct device *dev, uint32_t *dev_config)
+{
+	struct i2c_numaker_data *data = dev->data;
+
+	if (!dev_config) {
+		return -EINVAL;
+	}
+
+	k_sem_take(&data->lock, K_FOREVER);
+	*dev_config = data->dev_config;
+	k_sem_give(&data->lock);
+
+	return 0;
+}
+
+static int i2c_numaker_transfer(const struct device *dev, struct i2c_msg *msgs,
+				 uint8_t num_msgs, uint16_t addr)
```

Comment:
```c
static int i2c_numaker_transfer(const struct device *dev, struct i2c_msg *msgs, 
                                uint8_t num_msgs, uint16_t addr)                
```

you prob want to check your editor to make it do it for you, there's few odds indentations around the file :-)

Reply:
Yes, it's fixed after adjust my editor tab size from 4 as 8.
Thanks of your advice.